sr_robot_lib/src/MST.cpp

```diff
@@ -11,130 +11,249 @@
+  if (index % 2 == 0)
   {
-    int start = index * 3 / 2;
-    if (index % 2 == 0)
-    {
-      int byte1 = buffer[start];
-      int byte2 = buffer[start + 1];
-      return byte1 << 4 | byte2 >> 4 & 0x0F;
-    }
-    int byte1 = (int8_t)(buffer[start] << 4);
+    int byte1 = buffer[start];
     int byte2 = buffer[start + 1];
-    return byte1 << 4 | byte2 & 0xFF;
+
+    return byte1 << 4 | byte2 >> 4 & 0x0F;
   }
+  // If requested index is odd, return the last 4 bits of the current byte, and the next byte
+  int byte1 = (int8_t)(buffer[start] << 4);
+  int byte2 = buffer[start + 1];
 
-  template<class StatusType, class CommandType>
-  void MST<StatusType, CommandType>::update(StatusType *status_data)
+  return byte1 << 4 | byte2 & 0xFF;
+}
+
+/**
+  * Extract the tactile specific data from the incoming EtherCAT status message from the Palm.
+  * And update the tactile data structure.
+  *
+  * @param status_data Is a pointer to an incoming EtherCAT status message from the Palm.
+  */
+template<class StatusType, class CommandType>
+void MST<StatusType, CommandType>::update(StatusType *status_data)
+{
+  // Returns a bit mask describing which fingers contain MST tactile sensors
+  int tactile_mask = static_cast<int16u>(status_data->tactile_data_valid);
+  for (unsigned int id_sensor = 0; id_sensor < this->nb_tactiles; ++id_sensor)
   {
-    int tactile_mask = static_cast<int16u>(status_data->tactile_data_valid);
-    for (unsigned int id_sensor = 0; id_sensor < this->nb_tactiles; ++id_sensor)
+    // Assert the received MST data type
+    switch (static_cast<int32u>(status_data->tactile_data_type))
     {
-      switch (static_cast<int32u>(status_data->tactile_data_type))
-      {
-        case TACTILE_SENSOR_TYPE_MST_MAGNETIC_INDUCTION:
-          if (sr_math_utils::is_bit_mask_index_true(tactile_mask, id_sensor))
+      case TACTILE_SENSOR_TYPE_MST_MAGNETIC_INDUCTION:
+        // Check if the current sensor (id_sensor) is present in the bit mask
+        if (sr_math_utils::is_bit_mask_index_true(tactile_mask, id_sensor))
+        {
+          for (uint8_t taxel_index = 0; taxel_index < NUMBER_OF_TAXELS; taxel_index++)
+          {
+            geometry_msgs::Point taxel_magnetic_data;
+            // Decode incoming tactile data for each taxel and channel (X, Y or Z)
+            // And build ROS message
+            taxel_magnetic_data.x = read12bits(status_data->tactile[id_sensor].string, taxel_index * 3);
+            taxel_magnetic_data.y = read12bits(status_data->tactile[id_sensor].string, taxel_index * 3 + 1);
+            taxel_magnetic_data.z = read12bits(status_data->tactile[id_sensor].string, taxel_index * 3 + 2);
+            sensor_data.tactiles[id_sensor].magnetic_data[taxel_index] = taxel_magnetic_data;
+          }
+          // If MST sensor has Status Check enabled, retrieve status from latest measurment
+          if (diagnostic_data[id_sensor].git_revision.back() == '1')
```

Comment:
I think we should treat this as a bit, not as a certain character. (even though curretly this code would work because we only one one bit of that byte).
Something like this pseudocode:
```
if (convert_hex_string_to_byte(diagnostic_data[id_sensor].git_revision[-2:]) & 0x01)
{
```
If you want to leave it as it is, please add an explanation that this is a trick that will not work if we decide to use more flags in that bytes.

Reply:
Fixed here: https://github.com/shadow-robot/sr-ros-interface-ethercat/pull/265/commits/73c3ad6b307927ab702ba5d373a7d41316458e7c